### ISSUE TITLE
Using qty input stepper in 9.5.0-beta1 does not trigger a qty field change

### ISSUE THREAD

**Issue by medavidallsop** [woothemes/woocommerce#53031]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When testing https://developer.woocommerce.com/2024/11/20/call-for-feedback-stepper-layout-in-add-to-cart-block/

In 9.5.0-beta1 with add-to-cart-with-options-stepper-layout enabled, using the plus or minus buttons does not trigger a change event on the quantity field.

This breaks custom functionality like this:

```
$( '#custom-add-to-cart-form-wrap' ).find( '.cart input[name="quantity"]' ).on( 'change', function() {

	someFunction();

});
```

In this example `someFunction()` will not get run.

### Expected behavior

Using the plus or minus in the stepper fires a change event when the quantity field gets changed.

### Actual behavior

No change event fired.

### Steps to reproduce

Setup the beta as per https://developer.woocommerce.com/2024/11/20/call-for-feedback-stepper-layout-in-add-to-cart-block/

Add this code to the product page, then use the stepper, no alert is shown.

```
$( '.cart input[name="quantity"]' ).on( 'change', function() {

	alert( 'change fired' );

});
```

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): https://wc95beta.s3-tastewp.com
Site address (URL): https://wc95beta.s3-tastewp.com
WC Version: 9.5.0-beta.1
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.8.2
Log Directory Writable: ✔
WP Version: 6.7
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: TasteWP-S3 Official/3.0.0
PHP Version: 8.0.30
PHP Post Max Size: 128 MB
PHP Time Limit: 100
PHP Max Input Vars: 5000
cURL Version: 7.68.0
OpenSSL/1.1.1f

SUHOSIN Installed: –
MySQL Version: 8.0.32-0ubuntu0.20.04.2
Max Upload Size: 60 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 9.4.2
WC Database Prefix: wp_
Total Database Size: 2.93MB
Database Data Size: 1.22MB
Database Index Size: 1.71MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wpfm_backup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 24
page: 7
post: 2
product: 18
product_variation: 7
wp_global_styles: 2
wp_navigation: 1
wp_template: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (2) ###

WooCommerce Beta Tester: by WooCommerce – 2.3.2
WooCommerce: by Automattic – 9.5.0-beta.1

### Dropin Plugins () ###

maintenance.php: maintenance.php

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 2 KB

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/ -  Contains the woocommerce/cart block
Checkout: #9 - /checkout/ -  Contains the woocommerce/checkout block
My account: #10 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.3
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
Theme type: Block theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /s3-wc95beta/wordpress/wp-content/plugins/woocommerce/templates/block-notices/error.php
/s3-wc95beta/wordpress/wp-content/plugins/woocommerce/templates/block-notices/notice.php
/s3-wc95beta/wordpress/wp-content/plugins/woocommerce/templates/block-notices/success.php

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store
add-to-cart-with-options-stepper-layout

Disabled Features: product-data-views
experimental-blocks
coming-soon-newsletter-template
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system
blueprint
reactify-classic-payments-settings
use-wp-horizon
beta-tester-slotfill-examples

Daily Cron: ✔ Next scheduled: 2024-11-22 09:25:00 +00:00
Options: ✔
Notes: 56
Onboarding: skipped

### Action Scheduler ###

Complete: 33
Oldest: 2024-11-21 09:26:03 +0000
Newest: 2024-11-21 09:31:02 +0000

Pending: 2
Oldest: 2024-11-21 09:40:15 +0000
Newest: 2024-11-22 09:26:03 +0000

Failed: 1
Oldest: 2024-11-21 09:26:03 +0000
Newest: 2024-11-21 09:26:03 +0000

### Status report information ###

Generated at: 2024-11-21 09:39:36 +00:00
`

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by helgatheviking** [comment]
Seconding! I have seen this in my own explorations as well. It can be resolved with 

```
const changeEvent = new Event( 'change', {
	bubbles: true,
	cancelable: true
});

// Dispatch a change event on the input element.
inputElement.dispatchEvent(changeEvent);
```

I've actually done this locally.... I will try to send a PR.

**Comment by gigitux** [comment]
Thanks for testing the stepper layout and reporting this issue!

According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event):

> The change event is fired for `<input>, <select>, and <textarea>` elements **when the user modifies the element's value**. Unlike the input event, **the change event is not necessarily fired for each alteration to an element's value.**

It seems correct that the event isn't fired when the user clicks on the stepper buttons. Instead, I suggest adding an event listener to the stepper buttons.

**Comment by medavidallsop** [comment]
> I suggest adding an event listener to the stepper buttons.

Sorry, by this do you mean in WooCommerce directly to fire the change event on those buttons being clicked, or do you mean a plugin that has some form of functionality based off quantity field changes will need to add additional code to still function as before?

**Comment by gigitux** [comment]
Plugins should revisit their logic, if needed, to improve compatibility with the new layout. This is one of the reasons for the call for feedback.

**Comment by medavidallsop** [comment]
Okay, in that case can we ensure the the IDs/classes of those elements never get changed!

**Comment by helgatheviking** [comment]
>Instead, I suggest adding an event listener to the stepper buttons.
>Plugins should revisit their logic

Could not disagree more. If core fires the change event, then lots of plugins will just keep working automagically.

I've had to deal with themes adding their own plus/minus buttons to the quantity inputs for *years*. When they don't trigger the change event manually (after programmatically _changing_ the value) then users have assumed my plugin was broken because it listens to to changes on the input itself.

This is a cross compatibility issue. If y'all do this from core itself then you can drastically simplify the plugin ecosystem developers's lives (read my life :) ). 

The plus/minus button changes the value of the input. IMO, it should therefore fire a change event.

**Comment by medavidallsop** [comment]
I second this, for me it means adding a few JS lines to a plugin - which I can absolutely do, but is every plugin developer going to do it, or even know they need to?

**Comment by gigitux** [comment]
Thanks for your valuable feedback! We decided to trigger the “change” event when the stepper buttons are clicked. This is the PR: https://github.com/woocommerce/woocommerce/pull/53343

Feel free to test it and provide further feedback! Thanks! 🙇

**Comment by medavidallsop** [comment]
@gigitux Thank you! This will eliminate a ton of headaches for third party developers, and also means we don't have to refactor anything to be compatible with this new feature.

**Comment by gigitux** [comment]
I just merged the PR. The change will be available in WooCommerce 9.6

**Comment by helgatheviking** [comment]
Thank you @gigitux This looks great!
